packages/ui/src/ui/pages/navigation/tabs/Flow/PipelineSpec/PipelineSpec.tsx

```diff
@@ -0,0 +1,218 @@
+import {getFlowSpecYsonSettings} from '../../../../../store/selectors/thor/unipika';
+import {FlowSpecState} from '../../../../../store/reducers/flow/specs';
+
+import Yson from '../../../../../components/Yson/Yson';
+import Icon from '../../../../../components/Icon/Icon';
+import {YTDFDialog, makeErrorFields} from '../../../../../components/Dialog/Dialog';
+import ErrorBlock from '../../../../../components/Block/Block';
+import Loader from '../../../../../components/Loader/Loader';
+import {UnipikaSettings} from '../../../../../components/Yson/StructuredYson/StructuredYsonTypes';
+
+import './PipelineSpec.scss';
+
+const block = cn('yt-pipeline-spec');
+
+type PipelineSpecProps = {
+    path: string;
+    error: YTError | undefined;
+    data: FlowSpecState['data'];
+    name: string;
+    onSave: (data: PipelineSpecProps['data']) => Promise<void>;
+};
+
+function PipelineSpec({path, data, error, name, onSave}: PipelineSpecProps) {
+    const [showEdit, setShowEdit] = React.useState(false);
+
+    const settings = useSelector(getFlowSpecYsonSettings);
+
+    return (
+        <React.Fragment>
+            {Boolean(error) && <ErrorBlock error={error} />}
+            <Yson
+                value={data}
+                settings={settings}
+                virtualized
+                folding
+                extraTools={
+                    <React.Fragment>
+                        <Button view="outlined" onClick={() => setShowEdit(true)}>
+                            <Icon awesome="pencil" />
+                            Edit {name}
+                        </Button>
+                    </React.Fragment>
+                }
+            />
+            <EditSpecDialog
+                path={path}
+                name={name}
+                visible={showEdit}
+                onSpecApply={(spec) => onSave({...data!, spec})}
+                onClose={() => setShowEdit(false)}
+                spec={data?.spec}
+                settings={settings}
+            />
+        </React.Fragment>
+    );
+}
+
+type FormValues = {
+    text: {value?: string};
+    path: string;
+};
+
+function EditSpecDialog({
+    path,
+    visible,
+    spec,
+    onClose,
+    onSpecApply,
+    name,
+    settings,
+}: Pick<PipelineSpecProps, 'name' | 'path'> & {
+    spec?: unknown;
+    visible: boolean;
+    onClose: () => void;
+    settings: UnipikaSettings;
+    onSpecApply: (spec: unknown) => Promise<void>;
+}) {
+    const [error, setError] = React.useState<YTError | undefined>();
+
+    const text = React.useMemo(() => {
+        return {value: JSON.stringify(spec, null, 4)};
+    }, [spec]);
+
+    return (
+        visible && (
+            <YTDFDialog<FormValues>
+                visible
+                size="l"
+                onClose={onClose}
+                initialValues={{text, path}}
+                headerProps={{
+                    title: `Edit ${name}`,
+                }}
+                onAdd={async (f) => {
```

Comment:
I would put this function under useCallback. Because each time the component is updated, a new onAdd function will be created, which may lead to undesirable effects, such as repeated calls to useEffect inside YTDFDialog.

Reply:
Such inplace definition of callbacks allows to use inference of types of parameters which pretty comfortable to write code this way. Also the form has only two fields, so it seems it will never have any performance issues. Lets leave it in current state.